- Select significance level
- Fit our model with all possible independent variables.
- Consider variable with highest p-value.
- If p-value is greater than significance level, remove variable
- Again fit the model without removed variable.

#### Backward Elimination For Multiple Linear Regression:

In [1]:
# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('salary matrix.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder = LabelEncoder()
X[:, 0] = labelencoder.fit_transform(X[:, 0])
onehotencoder = OneHotEncoder(categorical_features = [0])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

# Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

C:\Users\dhruv.aggarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


as we know in multiple linear regression,

$ y = b0+b1X1+b2X2+b3X3+….+bnXn $

we can also represent it as

$ y = b0X0+b1X1+b2X2+b3X3+….+bnXn $
where $ X0 = 1 $

So we can add one column with all values as 1 to represent $ b0X0 $.

In [3]:
import statsmodels.formula.api as sm
X = np.append ( arr = np.ones([30,1]).astype(int), values = X, axis = 1)

This is done because statsmodels library requires it to be done for constants.

Now, according to backward elimination for multiple linear regression algorithm, let us fit all variables in our model

In [4]:
X_opt = X[:,[0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.961
Model:                            OLS   Adj. R-squared:                  0.952
Method:                 Least Squares   F-statistic:                     117.2
Date:                Sun, 03 Mar 2019   Prob (F-statistic):           4.71e-16
Time:                        16:09:30   Log-Likelihood:                -300.09
No. Observations:                  30   AIC:                             612.2
Df Residuals:                      24   BIC:                             620.6
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.146e+04   4689.376      4.576      0.000    1.18e+04    3.11e+04
x1         -1421.4255   2683.353     -0.530      0.601   -6959.595    4116.744
x2            92.8656   2735.524      0.034      0.973   -5552.978    5738.710
x3             3.3361      2.410      1.384      0.179      -1.637       8.310
x4          -423.7607   1282.271     -0.330      0.744   -3070.237    2222.716
x5          9437.2530    510.978     18.469      0.000    8382.645    1.05e+04
==============================================================================
Omnibus:                        2.020   Durbin-Watson:                   1.918
Prob(Omnibus):                  0.364   Jarque-Bera (JB):                1.629
Skew:                           0.414   Prob(JB):                        0.443
Kurtosis:                       2.214   Cond. No.                     8.00e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large,  8e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Now  we have to remove 2nd column i.e. x2 because its p-value is maximum i.e. 0.973 ,

In [5]:
X_opt = X[:,[0,1,3,4,5]]
X_opt

array([[1.000e+00, 0.000e+00, 2.300e+03, 0.000e+00, 1.100e+00],
       [1.000e+00, 1.000e+00, 2.100e+03, 1.000e+00, 1.300e+00],
       [1.000e+00, 0.000e+00, 2.104e+03, 2.000e+00, 1.500e+00],
       [1.000e+00, 0.000e+00, 1.200e+03, 1.000e+00, 2.000e+00],
       [1.000e+00, 1.000e+00, 1.254e+03, 2.000e+00, 2.200e+00],
       [1.000e+00, 0.000e+00, 1.236e+03, 1.000e+00, 2.900e+00],
       [1.000e+00, 0.000e+00, 1.452e+03, 2.000e+00, 3.000e+00],
       [1.000e+00, 1.000e+00, 1.789e+03, 1.000e+00, 3.200e+00],
       [1.000e+00, 0.000e+00, 1.645e+03, 1.000e+00, 3.200e+00],
       [1.000e+00, 0.000e+00, 1.258e+03, 0.000e+00, 3.700e+00],
       [1.000e+00, 1.000e+00, 1.478e+03, 3.000e+00, 3.900e+00],
       [1.000e+00, 0.000e+00, 1.257e+03, 2.000e+00, 4.000e+00],
       [1.000e+00, 0.000e+00, 1.596e+03, 1.000e+00, 4.000e+00],
       [1.000e+00, 1.000e+00, 1.256e+03, 2.000e+00, 4.100e+00],
       [1.000e+00, 0.000e+00, 1.489e+03, 3.000e+00, 4.500e+00],
       [1.000e+00, 0.000e+00, 1.236e+03,

Again fit our model after removing 2nd column which was a dummy variable column


In [6]:
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.961
Model:                            OLS   Adj. R-squared:                  0.954
Method:                 Least Squares   F-statistic:                     152.6
Date:                Sun, 03 Mar 2019   Prob (F-statistic):           3.54e-17
Time:                        16:13:08   Log-Likelihood:                -300.09
No. Observations:                  30   AIC:                             610.2
Df Residuals:                      25   BIC:                             617.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.151e+04   4315.092      4.986      0.000    1.26e+04    3.04e+04
x1         -1466.4735   2285.214     -0.642      0.527   -6172.961    3240.014
x2             3.3230      2.331      1.426      0.166      -1.477       8.124
x3          -422.1845   1255.570     -0.336      0.739   -3008.079    2163.710
x4          9439.2110    497.467     18.975      0.000    8414.659    1.05e+04
==============================================================================
Omnibus:                        2.056   Durbin-Watson:                   1.918
Prob(Omnibus):                  0.358   Jarque-Bera (JB):                1.630
Skew:                           0.407   Prob(JB):                        0.443
Kurtosis:                       2.200   Cond. No.                     7.20e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.2e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Value for x3 is highest i.e. 0.739 hence we will remove it. 
This means, x3 column, which was number of certifications done, was not having significant impact on salary prediction.

Execute below line to remove x3

In [8]:
X_opt = X[:,[0,1,3,5]]
X_opt

array([[1.000e+00, 0.000e+00, 2.300e+03, 1.100e+00],
       [1.000e+00, 1.000e+00, 2.100e+03, 1.300e+00],
       [1.000e+00, 0.000e+00, 2.104e+03, 1.500e+00],
       [1.000e+00, 0.000e+00, 1.200e+03, 2.000e+00],
       [1.000e+00, 1.000e+00, 1.254e+03, 2.200e+00],
       [1.000e+00, 0.000e+00, 1.236e+03, 2.900e+00],
       [1.000e+00, 0.000e+00, 1.452e+03, 3.000e+00],
       [1.000e+00, 1.000e+00, 1.789e+03, 3.200e+00],
       [1.000e+00, 0.000e+00, 1.645e+03, 3.200e+00],
       [1.000e+00, 0.000e+00, 1.258e+03, 3.700e+00],
       [1.000e+00, 1.000e+00, 1.478e+03, 3.900e+00],
       [1.000e+00, 0.000e+00, 1.257e+03, 4.000e+00],
       [1.000e+00, 0.000e+00, 1.596e+03, 4.000e+00],
       [1.000e+00, 1.000e+00, 1.256e+03, 4.100e+00],
       [1.000e+00, 0.000e+00, 1.489e+03, 4.500e+00],
       [1.000e+00, 0.000e+00, 1.236e+03, 4.900e+00],
       [1.000e+00, 1.000e+00, 2.311e+03, 5.100e+00],
       [1.000e+00, 0.000e+00, 2.245e+03, 5.300e+00],
       [1.000e+00, 0.000e+00, 2.365e+03, 5.900

In [9]:
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.956
Method:                 Least Squares   F-statistic:                     210.7
Date:                Sun, 03 Mar 2019   Prob (F-statistic):           2.35e-18
Time:                        16:15:35   Log-Likelihood:                -300.16
No. Observations:                  30   AIC:                             608.3
Df Residuals:                      26   BIC:                             613.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.122e+04   4154.694      5.108      0.000    1.27e+04    2.98e+04
x1         -1479.8541   2245.558     -0.659      0.516   -6095.665    3135.956
x2             3.3059      2.290      1.443      0.161      -1.402       8.014
x3          9336.1224    385.024     24.248      0.000    8544.694    1.01e+04
==============================================================================
Omnibus:                        2.340   Durbin-Watson:                   1.842
Prob(Omnibus):                  0.310   Jarque-Bera (JB):                1.662
Skew:                           0.375   Prob(JB):                        0.436
Kurtosis:                       2.124   Cond. No.                     7.04e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.04e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Now, here p value of x1 i.e. another dummy variable from our perspective is 0.516 which needs to be removed. So our X_opt should be

In [10]:
X_opt = X[:,[0,3,5]]
X_opt

array([[1.000e+00, 2.300e+03, 1.100e+00],
       [1.000e+00, 2.100e+03, 1.300e+00],
       [1.000e+00, 2.104e+03, 1.500e+00],
       [1.000e+00, 1.200e+03, 2.000e+00],
       [1.000e+00, 1.254e+03, 2.200e+00],
       [1.000e+00, 1.236e+03, 2.900e+00],
       [1.000e+00, 1.452e+03, 3.000e+00],
       [1.000e+00, 1.789e+03, 3.200e+00],
       [1.000e+00, 1.645e+03, 3.200e+00],
       [1.000e+00, 1.258e+03, 3.700e+00],
       [1.000e+00, 1.478e+03, 3.900e+00],
       [1.000e+00, 1.257e+03, 4.000e+00],
       [1.000e+00, 1.596e+03, 4.000e+00],
       [1.000e+00, 1.256e+03, 4.100e+00],
       [1.000e+00, 1.489e+03, 4.500e+00],
       [1.000e+00, 1.236e+03, 4.900e+00],
       [1.000e+00, 2.311e+03, 5.100e+00],
       [1.000e+00, 2.245e+03, 5.300e+00],
       [1.000e+00, 2.365e+03, 5.900e+00],
       [1.000e+00, 1.500e+03, 6.000e+00],
       [1.000e+00, 1.456e+03, 6.800e+00],
       [1.000e+00, 1.760e+03, 7.100e+00],
       [1.000e+00, 2.400e+03, 7.900e+00],
       [1.000e+00, 2.148e+03, 8.20

In [11]:
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     322.5
Date:                Sun, 03 Mar 2019   Prob (F-statistic):           1.42e-19
Time:                        16:18:13   Log-Likelihood:                -300.41
No. Observations:                  30   AIC:                             606.8
Df Residuals:                      27   BIC:                             611.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.102e+04   4099.682      5.127      0.000    1.26e+04    2.94e+04
x1             3.1236      2.250      1.389      0.176      -1.492       7.739
x2          9340.2046    380.920     24.520      0.000    8558.621    1.01e+04
==============================================================================
Omnibus:                        1.451   Durbin-Watson:                   1.842
Prob(Omnibus):                  0.484   Jarque-Bera (JB):                1.323
Skew:                           0.388   Prob(JB):                        0.516
Kurtosis:                       2.325   Cond. No.                     7.01e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.01e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Now remove x1 column which represents our worked hours, p values is coming as 0.176 which is way over our significance level 0.05 hence we can remove x1 i.e. worked hours column too.

In [12]:
X_opt = X[:,[0,5]]

- we now have only two columns left and if you closely watch p values for const and x2 in above p value table, they are 0.000 which are below significance level 0.05.
- It means they are important and have impact on salary predictions. 
- So our backward elimination for multiple linear regression is now stopped. 
- We will predict salaries based on current state of X_opt.

#### Predict salaries for X_opt

In [14]:
# Splitting the dataset into the Training set and Test set
X_opt_train, X_opt_test, y_opt_train, y_opt_test = train_test_split(X_opt, y, test_size = 1/3, random_state = 0)
regressor_opt = LinearRegression()
regressor_opt.fit(X_opt_train, y_opt_train)
 
y_opt_pred = regressor_opt.predict(X_opt_test)

In [15]:
y_opt_pred

array([ 40835.10590871, 123079.39940819,  65134.55626083,  63265.36777221,
       115602.64545369, 108125.8914992 , 116537.23969801,  64199.96201652,
        76349.68719258, 100649.1375447 ])

predictions obtained from backward elimination for multiple linear regression are much closer to actual results and same as simple linear regression in our case.